Transfer learning example

jbschiratti opened the pull request:
## What does this PR do?
Addresses issue #514. Following up on [this discussion](https://github.com/PyTorchLightning/pytorch-lightning/issues/514#issuecomment-610898707), this PR proposes to add a self-contained example which shows how a pretrained network (such as ResNet50) can be fine tuned within a LightningModule. 

## PR review    
Anyone in the community is free to review the PR 🙂

hcjghr commented:
Hi @jbschiratti 

Thanks for such a nice example. I'm rather new in the field so hopefully my question will not be too off base. I was just going over the code and I noticed that in your example the BatchNorm layers will always remain in training mode (as train_bn is always set to self.hparams.train_bn when calling the freeze function). That is even when performing validation or evaluation. I understand the code potentially allows for the BN layers to be set to eval (if the train_bn=False) but I am just wondering if there is a specific reason why do you always leave BN in train mode? Why not have them train in the training stage and eval in the validation/testing?

Just to clerify I'm not arguing it should be different, I'm just asking for the reasoning behind it.

Borda commented:
> Thanks for such a nice example. I'm rather new in the field so hopefully my question will not be too off base. I was just going over the code and I noticed ...

Thank you for your interest and help with this addition, may you please use review tab of this PR to write your comments directly to the sections you are talking about... it will make the discussion clearer and a bit more concrete :]

jbschiratti commented on pl_examples/domain_templates/computer_vision_fine_tuning.py:
```diff
@@ -0,0 +1,405 @@
+        return self.loss_func(input=logits, target=labels)
+
+    def train(self, mode=True):
+        super(TransferLearningModel, self).train(mode=mode)
```
Fixed.

jbschiratti commented on pl_examples/domain_templates/computer_vision_fine_tuning.py:
```diff
@@ -0,0 +1,405 @@
+                                milestones=self.hparams.milestones,
+                                gamma=self.hparams.lr_scheduler_gamma)
+
+        return [optimizer], [scheduler]
```
I am not sure what should be fixed here. The return types are OK here, are they?

Borda replied:
there shall be a tuple of 4

jbschiratti replied:
Sorry, but I still don't get it. Why a 4-tuple?
In [this basic example](https://github.com/PyTorchLightning/pytorch-lightning/blob/990fd22488f81dd38b1000934e5ab83cab6170f2/pl_examples/models/lightning_template.py#L181), a 2-tuple is returned. And the [doc of `configure_optimizers`](https://github.com/PyTorchLightning/pytorch-lightning/blob/990fd22488f81dd38b1000934e5ab83cab6170f2/pytorch_lightning/core/lightning.py#L994) allows to return two lists (one for optimizers, one for LR schedulers). Should the doc be updated or am I missing something?

awaelchli replied:
Yes it is good like this. The Union is a choice between the four options, and here the fourth one applies, which is a tuple. I don't see anything wrong with it :)

jbschiratti commented:
@hcjghr Thank you for spotting this. It's was a bug! The way I see it, in the evaluation loop when `model.eval()` is called, the BatchNorm layers (as well as the other layers) should be in eval mode (`training=False`). This is what I had in mind and it is now fixed.

@Borda I fixed the docstrings and added `argparse`. Thank you for the comments!

Borda commented:
pls add note to changelog :rabbit:

hcjghr commented on pl_examples/domain_templates/computer_vision_fine_tuning.py:
```diff
@@ -0,0 +1,441 @@
+                    yield param
+    else:
+        for child in children:
+            filter_params(module=child, train_bn=train_bn)
```
I think this line should actually be: 
`yield from filter_params(module=child, train_bn=train_bn)`

This way the yielded parameters from a nested call actually get back to the _optimizer.add_param_group_ otherwise (as it is now) they just get "lost" before being returned to the initial call.

jbschiratti replied:
Thank you for pointing this out ! I missed it. It's now fixed.

awaelchli commented on pl_examples/domain_templates/computer_vision_fine_tuning.py:
```diff
@@ -0,0 +1,447 @@
+            val_loss = output['val_loss']
+            val_acc = output['num_correct']
+            # reduce manually when using dp
+            if self.trainer.use_dp or self.trainer.use_ddp2:
```
can you not just do `torch.stack([output[metric] for output in outputs]).mean() `?
Then you include all cases, dp, ddp, ... without all of this code. same applies to the training_epoch_end.

jbschiratti replied:
Done!

awaelchli replied:
I think now you divide by the length twice in case of val_loss (you have the mean above already)

jbschiratti replied:
Yes, of course! Thanks! It's fixed.

jbschiratti commented on pl_examples/domain_templates/computer_vision_fine_tuning.py:
```diff
@@ -0,0 +1,447 @@
+        normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],
+                                         std=[0.229, 0.224, 0.225])
+
+        train_dataset = ImageFolder(root=data_path.joinpath('train'),
```
You're right, I forgot about the `prepare_data` method.

jbschiratti commented on pl_examples/domain_templates/computer_vision_fine_tuning.py:
```diff
@@ -0,0 +1,447 @@
+        return self.__dataloader(train=False)
+
+    @staticmethod
+    def add_specific_args(parent_parser):
```
Fixed!

jbschiratti commented on pl_examples/domain_templates/computer_vision_fine_tuning.py:
```diff
@@ -0,0 +1,447 @@
+    """Yield the trainable parameters of a given module.
+
+    Args:
+        module (torch.nn.Module): A given module
```
Fixed!

jbschiratti commented:
Thanks @awaelchli for the review and the comments!

awaelchli commented on pl_examples/domain_templates/computer_vision_fine_tuning.py:
```diff
@@ -0,0 +1,440 @@
+
+        # 2. Compute loss & accuracy:
+        val_loss = self.loss(y_true, y_logits)
+        num_correct = torch.eq(y_bin.view(-1), y_true.view(-1)).sum().item()
```
```suggestion
        num_correct = torch.eq(y_bin.view(-1), y_true.view(-1)).sum()
```
Sorry to be pedantic, but it gives an error now that we are using torch.stack in epoch end, because stacking a list of ints doesn't work. Let's keep it a tensor here? same applies in training.

awaelchli commented on pl_examples/domain_templates/computer_vision_fine_tuning.py:
```diff
@@ -0,0 +1,440 @@
+
+        # 2. Compute loss & accuracy:
+        train_loss = self.loss(y_true, y_logits)
+        num_correct = torch.eq(y_bin.view(-1), y_true.view(-1)).sum().item()
```
```suggestion
        num_correct = torch.eq(y_bin.view(-1), y_true.view(-1)).sum()
```

awaelchli commented:
I noticed that the downloaded dataset is not ignored in version control. Could we maybe move it to a subfolder datasets and add a .gitignore in domain templates folder?

jbschiratti commented:
This is strange because the context manager 

```python
with TemporaryDirectory(dir=hparams.root_data_path) as tmp_dir:
    ...
```

should delete the temporary folder in which the data is downloaded.

awaelchli commented:
ah ok, so it is also supposed to do that on keyboard interrupt? Maybe it's because I'm on Windows currently.

jbschiratti commented:
I tried to stop the script with CTRL+C during the 1st epoch and the temporary folder was deleted (on Linux). But I cannot guarantee this always works.

awaelchli commented:
Can now also confirm it works fine on Linux, so it's just a Windows thing, so I guess we can keep it like that.

Borda commented on pl_examples/domain_templates/computer_vision_fine_tuning.py:
```diff
@@ -0,0 +1,440 @@
+    """Unfreezes a module and adds its parameters to an optimizer."""
+    _make_trainable(module)
+    params_lr = optimizer.param_groups[0]['lr'] if lr is None else float(lr)
+    optimizer.add_param_group(
+        {'params': filter_params(module=module, train_bn=train_bn),
+         'lr': params_lr / 10.,
+         })
```
```suggestion
    optimizer.add_param_group({
        'params': filter_params(module=module, train_bn=train_bn),
        'lr': params_lr / 10.,
    })
```

jbschiratti replied:
Done!

Borda commented on pl_examples/domain_templates/computer_vision_fine_tuning.py:
```diff
@@ -0,0 +1,440 @@
+    def loss(self, labels, logits):
+        return self.loss_func(input=logits, target=labels)
+
+    def train(self, mode=True):
```
what is the `mode` for? could it be more descriptive?

jbschiratti replied:
See https://github.com/pytorch/pytorch/blob/d37a4861b8a5eed3d9a1340484d1efb0f48aa59e/torch/nn/modules/module.py#L1067. This line overrides the `train` method of the Pytorch module. I will add a docstring specifying what `mode` does.

Borda replied:
you are right, we may rename it...
doe you have suggestion about a better name? @PyTorchLightning/core-contributors

jbschiratti replied:
I am not sure we can rename it. In the [evaluation loop (L330)](https://github.com/PyTorchLightning/pytorch-lightning/blob/42d5cfc3b056b4c82a77a7cdcb8eafc63a812b67/pytorch_lightning/trainer/evaluation_loop.py#L330), `model.train()` is called and here, `model` refers (if I am not mistaken) to the LightningModule. We want to override this `train` method so that, at the end of the evaluation loop, some parameters (in specific layers) remain frozen (that is, with `requires_grad=False`) if needed.

Borda commented on pl_examples/domain_templates/computer_vision_fine_tuning.py:
```diff
@@ -0,0 +1,440 @@
+        num_correct = torch.eq(y_bin.view(-1), y_true.view(-1)).sum()
+
+        # 3. Outputs:
+        tqdm_dict = {'train_loss': train_loss}
```
no need to name it after particular progress bar lib...

jbschiratti replied:
Done!

Borda commented on pl_examples/domain_templates/computer_vision_fine_tuning.py:
```diff
@@ -0,0 +1,440 @@
+    @staticmethod
+    def add_model_specific_args(parent_parser):
+        parser = argparse.ArgumentParser(parents=[parent_parser])
+        parser.add_argument('--backbone',
```
use `add_argparse_args` so we limit duplication and add just the new/needed for a model

jbschiratti replied:
by "limit code duplication", you want me to remove [this line](https://github.com/PyTorchLightning/pytorch-lightning/pull/1564/files/ff19e3763919c38595731086ed8cd5a197b12ee8#diff-78fcc134908f8b115ba9b07e9fc99732R427)?

Borda replied:
I mean remove lines which are generated from Trainer arguments... does it make sense?

Borda replied:
https://github.com/PyTorchLightning/pytorch-lightning/blob/9b86aea98bdaa73bc3bf8841f4dc794f46a3f2ac/pytorch_lightning/trainer/trainer.py#L589

Borda commented on pl_examples/domain_templates/computer_vision_fine_tuning.py:
```diff
@@ -0,0 +1,440 @@
+        to a temporary directory.
+    """
+
+    with TemporaryDirectory(dir=hparams.root_data_path) as tmp_dir:
```
I guess we want to keep the output folder

Borda replied:
@jbschiratti ^^

jbschiratti replied:
The folder in which the data was downloaded is deleted after the experiment. If you think we should leave the data untouched after the example has run, I can make another PR to fix this  :-)

williamFalcon commented:
@jbschiratti this is super cool.
Why don't we move this to https://github.com/PyTorchLightning/pytorch-lightning-bolts?

Borda commented:
I would keep it here in examples....

jbschiratti commented:
Thank you @williamFalcon 👍
